udev/net: introduce WakeOnLanSecureOnPassword=

yuwata opened the pull request:
Closes #20913.

poettering commented on man/systemd.link.xml:
```diff
@@ -548,6 +548,15 @@
+          <para>Specifies the SecureOn(tm) password for MagicPacket(tm). Takes 6 bytes in Ethernet MAC
+          hex format (<literal>aa:bb:cc:dd:ee:ff</literal>). This will be ignored when
+          <varname>WakeOnLan=</varname> is specified without <literal>secureon</literal>. Defaults to
+          unset, and the current value will not be changed.</para>
```
it's so weird that they use MAC addresses as passwords.

yuwata replied:
Slightly rephrased the explanation.

poettering commented on man/systemd.link.xml:
```diff
@@ -548,6 +548,15 @@
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>WakeOnLanSecureOnPassword=</varname></term>
```
i'd probably just name this "WakeOnLanPassword=". i.e. no need to repeat the trademark here, the generic naming as just "password" is clear enough.

poettering commented on man/systemd.link.xml:
```diff
@@ -548,6 +548,15 @@
+          <varname>WakeOnLan=</varname> is specified without <literal>secureon</literal>. Defaults to
+          unset, and the current value will not be changed.</para>
+        </listitem>
+      </varlistentry>
```
I think we should also make sure that if someone sets WakeOnLanPassword= but doesn't touch WakeOnLan= at all that we imply WakeOnLan=secureon.

poettering reviewed (changes requested):
looks great. 

(I think eventually we should make this work a bit more like the wireguard key handling, i.e. allow specifiying a path in the fs to read the password from, and support AF_UNIX sockets for that too, i.e. wireguard_read_key_file())

bluca commented on src/shared/ethtool-util.c:
```diff
@@ -1359,6 +1374,48 @@ int config_parse_wol(
+
+        assert(filename);
+        assert(rvalue);
+
```
`assert(password)`

rnhmjoj commented:
> (I think eventually we should make this work a bit more like the wireguard key handling, i.e. allow specifiying a path in the fs to read the password from

I was going to ask the same thing: it's better not to mix secrets and configuration, if possible. This is particularly probematic in NixOS where the (generated) configuration files are world-readable by default.

Anyway, I'll try test this patch as soon as I have time and report back. 
Thank you for working on this, @yuwata.

poettering commented:
(I think eventually we should also hook up the creds logic to wireguard and secureon, i.e. so that you can do LoadCredentials= in the networkd unit file to set these, or even inherit things down from nspawn params, and so on. i.e. try read_credential() on some name derived from the network config file name or so, and mayba single fallback to a fixed string)

yuwata commented:
@poettering, @bluca, and @rnhmjoj Thank you for your review and comments. Updated.
- Renamed WakeOnLanSecureOnPassword= -> WakeOnLanPassword=
- The setting now also support a boolean string and a path to a file which contains the password.

PTAL.

(I will address wireguard thing in other later PR.)

yuwata commented:
Oh, udevd is coredumped on focal-s390x. That may be related. I will rework on that.

yuwata commented:
I hope now the segfault is fixed.

poettering commented on man/systemd.link.xml:
```diff
@@ -537,8 +537,8 @@
               <listitem>
-                <para>Enable secureon(tm) password for MagicPacket(tm).
-                </para>
+                <para>Enable SecureOn(tm) password for MagicPacket(tm). Implied when
```
I am having trouble finding any spec for these techs. All I found is this:

https://www.amd.com/system/files/TechDocs/20213.pdf

And that doc never puts a TM or so on the MagicPacket thing. SecureOn is not mentioned at all there.

poettering replied:
(i'd probably just drop the TM though)

poettering commented on src/udev/net/link-config.c:
```diff
@@ -834,6 +867,14 @@ int config_parse_wol_password(
+                return 0;
+        }
+
+        r = parse_boolean(rvalue);
```
hmm, this is definitely a workable approach, but i wonder if it wouldn't be simpler and more natural to simply say that if `WakeOnLan=secureon` is selected but `WakeOnLanPassword=` is not we'll search for the password in the creds? And then not have any boolean parsing here at all?

That way: 

`WakeOnLan=secureon` without `WakeOnLanPassword=` → search for the password in the creds and use it if it exists
`WakeOnLanPassword=` without `WakeOnLan=` → imply `WakeOnLan=secureon`

poettering commented on src/shared/ethtool-util.c:
```diff
@@ -425,16 +431,37 @@ int ethtool_set_wol(int *ethtool_fd, const char *ifname, uint32_t wolopts) {
+                r = -errno;
 
-        return 0;
+        zero(ecmd);
```
Should use explicit_bzero_safe(), zero() is not enough compiler might optimize that away

poettering reviewed (changes requested):
looks excellent!
